util/build/sgx/src/libraries.rs

```diff
@@ -61,3 +81,59 @@ impl SgxLibraryCollection for [Library] {
         }
     }
 }
+
+/// Provides the necessary linker flags to link SGX libraries to the crate being
+/// built.
+///
+/// # Arguments
+///
+/// * `sgx` - The SGX environment for the current build.  This helps to
+///   determine if one needs to link to sim or hw libraries.
+pub fn link_to_sgx_libraries(sgx: &SgxEnvironment) -> Result<(), Error> {
+    let mut config = Config::new();
+    config
+        .exactly_version(SGX_VERSION)
+        .print_system_libs(true)
+        .cargo_metadata(false)
+        .env_metadata(true);
+    let lib_paths = if sgx.sgx_mode() == SgxMode::Simulation {
+        SGX_SIMULATION_LIBS
+    } else {
+        SGX_LIBS
+    }
+    .iter()
+    .map(|libname| Ok(config.probe(libname)?.link_paths))
+    .collect::<Result<Vec<Vec<PathBuf>>, PkgConfigError>>()?
+    .into_iter()
+    .flatten()
+    .collect::<HashSet<PathBuf>>();
+
+    for path in lib_paths {
+        if sgx.sgx_mode() == SgxMode::Simulation {
+            let cve_load = path.join("cve_2020_0551_load");
+            if cve_load.exists() {
+                rustc_link_search!(cve_load.display());
+            }
+        }
+        rustc_link_search!(path.display());
+    }
+
+    let sim_postfix = match sgx.sgx_mode() {
+        SgxMode::Hardware => "",
+        SgxMode::Simulation => "_sim",
+    };
+
+    // These need to be linked after the rest of the code so can't use the
+    // `rustc_lib_arg`
+    rustc_link_arg!(
+        "--whole-archive",
+        &format!("-lsgx_trts{}", sim_postfix),
+        "--no-whole-archive",
+        "-lsgx_tcxx",
+        "-lsgx_tcrypto",
+        &format!("-lsgx_tservice{}", sim_postfix),
+        "-lsgx_tstdc"
+    );
```

Comment:
Is the issue with this in the JSON the result of supporting simulation mode?

Reply:
Yes, the need to link to the `_sim` version of the libraries has kept this linking out of the json, for now.

I know we discussed the possibilities of separate json files for this use case.  Currently the json file is chosen via the config file in `.config/cargo.toml`.  Looking at the [config spec](https://doc.rust-lang.org/cargo/reference/config.html), I'm not seeing a way to read arbitrary environment variables to condition which json file to use.
This means that we either:
* Use the build scripts as currently is done
* pass the json file via build scripts with the [`CARGO_BUILD_TARGET`](https://doc.rust-lang.org/cargo/reference/config.html#buildtarget).  This means tools like `clippy` wouldn't know to use the json file.
* See if we can leverage [per-package-target](https://doc.rust-lang.org/cargo/reference/unstable.html#per-package-target) to conditionally choose the json.

At this time, I left the `_sim` versus non `_sim` logic in the build scripts as it gets us moving forward with the spec file.